codegen/smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/rendering/PaginatorGenerator.kt

```diff
@@ -141,18 +145,27 @@ class PaginatorGenerator : KotlinIntegration {
             ) {
                 withBlock("#T {", "}", ExternalTypes.KotlinxCoroutines.FlowGenerator) {
                     write("var cursor: #F = null", cursorSymbol)
-                    write("var isFirstPage: Boolean = true")
+                    write("var hasNextPage: Boolean = true")
                     write("")
-                    withBlock("while (isFirstPage || (cursor?.isNotEmpty() == true)) {", "}") {
+                    withBlock("while (hasNextPage) {", "}") {
                         withBlock("val req = initialRequest.copy {", "}") {
                             write("this.$markerLiteral = cursor")
                         }
                         write(
                             "val result = this@#1LPaginated.#1L(req)",
                             operationShape.defaultName(),
                         )
-                        write("isFirstPage = false")
                         write("cursor = result.$nextMarkerLiteral")
+
+                        val truncationMember = outputShape.members().singleOrNull {
+                            it.hasTrait(PaginationTruncationMember.ID)
+                        }
+                        val hasNextPageCondition = when (truncationMember) {
+                            null -> "cursor?.isNotEmpty() == true"
+                            else -> "result.${truncationMember.defaultName()} == true"
+                        }
```

Comment:
nit: consider `val hasNextPageCondition = truncationMember?.let { "result.${it.defaultName()}" } ?: "cursor?.isNotEmpty() == true"`

Reply:
Oh yep and I can extract the `== true` part as well. Good eye.